### ISSUE TITLE
Only pass in string literals into translation functions

### ISSUE THREAD

**Issue by rebeccahum** [Automattic/VIP-Coding-Standards#237]
## What problem would the enhancement address for VIP?

The WordPress translation functions `esc_attr_e()` and `esc_html_e()` are meant to be passed a string literal to be translated, escape it for display in an HTML attribute or text, and then echo it. If you pass these functions a variable instead of a quoted string, they won’t do any translation, so you should never use PHP variables of *any* kind inside a translation function’s strings. 

## What code should be reported as a violation?

```php
esc_html_e( $back_to_linked_post['text'] );
esc_attr_e( $this->plugin_name );
$bad = __($string, 'plugin-domain');
$bad =  __("You have $number tacos", 'plugin-domain');
$bad =  __('You have 12 tacos', $plugin_domain);
$bad = __('You have 12 tacos', PLUGIN_DOMAIN);
```

## What could should *not* be reported as a violation?

```php
esc_html_e( "Literal strings only." );
```

## Additional context
Further reading: http://ottopress.com/2012/internationalization-youre-probably-doing-it-wrong/

**Comment by GaryJones** [comment]
> I love the blurb!

Prompted by the new [Issue Templates](https://github.com/Automattic/VIP-Coding-Standards/tree/master/.github/ISSUE_TEMPLATE)!

Here's a considerably more recent and more rounded post on the subject, which I helped to edit:

https://pascalbirchler.com/text-domain-wordpress-internationalization/

And that explains why...

> Most people will want to store their plugin-domain in a constant or something of the like and I don't think that's too problematic

...will be problematic at some point in the future :-)

BTW, WPCS already has this covered in it's [I18n](https://github.com/WordPress-Coding-Standards/WordPress-Coding-Standards/blob/develop/WordPress/Sniffs/WP/I18nSniff.php) sniff - just search for "literal" and find the references.

Even if WPCS didn't have this though, I'd question if this is something that VIPCS really needs to highlight? Other than strings not being translated when they should be, what is the negative effect on the VIP-Go platform about someone trying to apply i18n incorrectly?

**Comment by rebeccahum** [comment]
> BTW, WPCS already has this covered in it's I18n sniff - just search for "literal" and find the references.

Oh, TIL!  Is this something we should add to the rulesets then?   

> Even if WPCS didn't have this though, I'd question if this is something that VIPCS really needs to highlight?

I guess that depends on the direction of VIPCS...what criteria is there towards the need for creating a sniff or adding one to the ruleset?

> Other than strings not being translated when they should be, what is the negative effect on the VIP-Go platform about someone trying to apply i18n incorrectly?

I'm not sure if there is other than the extra calling of `translate()`.

**Comment by GaryJones** [comment]
> I guess that depends on the direction of VIPCS...what criteria is there towards the need for creating a sniff or adding one to the ruleset?

May need a wider discussion, but if there's a sniff out there that already does what's needed, it seems a bad idea to spend the time creating and maintaining our own sniff.

I think VIPCS needs to cover, either through sniffs in VIPCS or sniffs elsewhere, all of the items in the documentation. While we might still have WPCS as a requirement, we shouldn't expect developers to have to run `--standard=WordPress-Extra` to find violations that they MUST fix. Sure, it will show a lot of items that probably SHOULD fix, but there's nothing in VIPCS that I'm aware of that actively conflicts with WPCS (i.e. tabs vs spaces). VIPCS is generally additive to WPCS for the things it cares about.



> Is this something we should add to the rulesets then?

Do the VIP Docs say that i18n MUST be done, and MUST be done correctly? Or even, saying it SHOULD be done? I can't see it in https://vip.wordpress.com/documentation/vip-go/code-review-blockers-warnings-notices/.

If there's a desire to have something like this anyway, then it may be worth considering an `WordPress-VIP-Go-Recommended` ruleset, which acts like `WordPress-Extra` does compared to `WordPress-Core`. I think right now, we can do without it, so long as the scope of VIPCS is contained to just what is (or should be) in the VIP Docs.
